### PR TITLE
DeviceManager/LVM: Synchronize all lvm calls

### PR DESCRIPTION
**Pull request by avalluri**
Often LVM stress tests are failing due to race conditions in executing LVM operations.
Hence, we added mutex locking so that only one LVM operation is in progress.
    
As part of this also made changes to cache LVM volumes by device manager so that it avoids repeated lvs calls.

This is an alternative approach to #334 to attack the lvm issues

### CODE REVIEW

File: pkg/pmem-device-manager/pmd-lvm.go [hunk 1]
```diff
@@ -112,6 +112,11 @@ func (lvm *pmemLvm) CreateDevice(name string, size uint64, nsmode string) error
 			if _, err := pmemexec.RunCommand("lvcreate", "-Zn", "-L", strSz, "-n", name, vg.name); err != nil {
 				glog.V(3).Infof("lvcreate failed with error: %v, trying for next free region", err)
 			} else {
+				// Sometimes needed softlinks that are supposed to create by udev are missing,
+				// hence we force create them by using vgscan
```
Comment (pohly) [inline pkg/pmem-device-manager/pmd-lvm.go]: "supposed to be created", missing dot at end of sentence.

Does this `vgscan` now race with udev? Where does this udev run - on the host or in the container?

I'm asking because we cannot know whether the host system has a running udev. I'm wondering whether we have a (so far) hidden dependency on the host system here.

**Comment by okartau** [comment]
not bad,  e2e in clear-govm cluster 10/10 OK, this has not happened for some time

File: pkg/pmem-device-manager/pmd-lvm.go (hunk 1 above)
Comment (avalluri) [reply pkg/pmem-device-manager/pmd-lvm.go]: I guess by defaults its using host udev, In my next commit I asked to disable udev syncing in container, so there should not be any race conditions with host udev.

**Comment by avalluri** [comment]
With this https://github.com/intel/pmem-csi/pull/336/commits/6740a95e01b275fefe39a17fbc4671f2a10e6197 change, seems like manual `vgscan` is not needed, Just asking lvm not to use `udev` should be enough to make softlinks in place.

**Comment by avalluri** [comment]
I run 14 attempts ithout any failure, Surprisingly I didn't see https://github.com/intel/pmem-csi/issues/337.  @okartau Can you try with this latest change, Lets see if its same results for you.

**Comment by pohly** [comment]
> /test

That doesn't work here. You need to log into Jenkins with your github account while connected to the Intel intranet. Then you can restart jobs there.

**Comment by pohly** [comment]
https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/organizations/jenkins/pmem-csi/detail/PR-336/5/pipeline was stuck (no worker?)

I tried to stop it, but without success. Force-push an amended commit and Jenkins should start a new job.

**Comment by avalluri** [comment]
> /test
> 
> That doesn't work here. You need to log into Jenkins with your github account while connected to the Intel intranet. Then you can restart jobs there.

I have tried it already, manually restarting the job, but there was no progress.

**Comment by okartau** [comment]
> I run 14 attempts without any failure, Surprisingly I didn't see #337.

That's likely because you try on qemu-based system. I can also run 10x e2e cleanly on a qemu-based systems. #337 happens on HW-based system only for me.

File: test/e2e/storage/sanity.go [hunk 2]
```diff
@@ -638,6 +652,12 @@ func (v volume) remove(vol *csi.Volume, volName string) {
 	framework.ExpectNoError(err, delete)
 }
 
+func skipIfNodeNotKnownWorker(nodeID string) {
```
Comment (pohly) [inline test/e2e/storage/sanity.go]: Instead of naming the function after how it is implemented, can you name it according to what it checks? We want to know whether that node can be restarted, so I think `canRestartNode` is better. It also matches the name `restartNode`, so it is obvious that the two are related.

**Comment by okartau** [comment]
"Fix unpublish target" commit was supposed to avoid the issue where unpublish is called with different path, leading to shred issue ?
I re-tested wit latest state of this PR, but there is still a volume that remains mounted and gets under shred risk.

**Comment by avalluri** [comment]
> "Fix unpublish target" commit was supposed to avoid the issue where unpublish is called with different path, leading to shred issue ?
> I re-tested wit latest state of this PR, but there is still a volume that remains mounted and gets under shred risk.

@okartau After checking logs you shared(on chat), I am pretty sure that those logs are not run using these lates changes, For example as per new code `stores state across reboots for single`  test should be skipped before creating any volume, but logs are showing differently:

` [It] stores state across reboots for single volume/home/okartau/go/src/github.com/intel/pmem-csi/test/e2e/storage/sanity.go:179 Jul 8 21:01:02.125: INFO: ExecWithOptions {Command:[sync] Namespace:default PodName:pmem-csi-node-testing-v5gdc ContainerName:socat Stdin: CaptureStdout:true CaptureStderr:true PreserveWhitespace:false} Jul 8 21:01:02.125: INFO: >>> kubeConfig: /home/okartau/kubeconfig [1mSTEP[0m: Exec Output: pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.365353 1 glog.go:58] GRPC call: /csi.v1.Controller/ListVolumes [1mSTEP[0m: state-volume: creating a single node writer volume pmem-csi-controller-0/pmem-driver: I0708 18:01:02.367643 1 glog.go:58] GRPC call: /csi.v1.Controller/CreateVolume pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.365379 1 glog.go:58] GRPC request: {} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.365991 1 glog.go:58] ListVolumes pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.366000 1 glog.go:58] GRPC response: {} pmem-csi-controller-0/pmem-driver: I0708 18:01:02.367683 1 glog.go:58] GRPC request: {"accessibility_requirements":{"preferred":[{"segments":{"pmem-csi.intel.com/node":"XXXXX"}}],"requisite":[{"segments":{"pmem-csi.intel.com/node":"XXXXX"}}]},"capacity_range":{"required_bytes":11534336},"name":"state-volume-8DA0F868-163CAD76","volume_capabilities":[{"AccessType":{"Mount":{}},"access_mode":{"mode":1}}]} pmem-csi-controller-0/pmem-driver: I0708 18:01:02.373368 1 glog.go:58] CreateVolume: Name: state-volume-8DA0F868-163CAD76 req.Required: 11534336 req.Limit: 0 pmem-csi-controller-0/pmem-driver: I0708 18:01:02.373393 1 glog.go:58] Connecting to node controller : tcp://NN.NNN.NN.NNN:10001 pmem-csi-controller-0/pmem-driver: I0708 18:01:02.373410 1 glog.go:58] Connecting to NN.NNN.NN.NNN:10001 pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.379525 1 glog.go:58] GRPC call: /csi.v1.Controller/CreateVolume pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.379566 1 glog.go:58] GRPC request: {"accessibility_requirements":{"preferred":[{"segments":{"pmem-csi.intel.com/node":"XXXXX"}}],"requisite":[{"segments":{"pmem-csi.intel.com/node":"XXXXX"}}]},"capacity_range":{"required_bytes":11534336},"name":"state-volume-8DA0F868-163CAD76","parameters":{"_id":"5926bc3d-a1aa-11e9-bb06-ae52abbc8405"},"volume_capabilities":[{"AccessType":{"Mount":{}},"access_mode":{"mode":1}}]} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.385804 1 glog.go:58] CreateVolume: Name: state-volume-8DA0F868-163CAD76 req.Required: 11534336 req.Limit: 0 pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.386502 1 glog.go:58] Executing: vgs --noheadings --nosuffix -o vg_name,vg_size,vg_free,vg_tags --units B ndbus0region1fsdax ndbus0region0fsdax pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.462265 1 glog.go:58] Output: ndbus0region0fsdax 266283778048 266283778048 fsdax pmem-csi-node-6crnt/pmem-driver: ndbus0region1fsdax 266283778048 266283778048 fsdax pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.462307 1 glog.go:58] Executing: lvcreate -Zn -L 11 -n 5926bc3d-a1aa-11e9-bb06-ae52abbc8405 ndbus0region0fsdax pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.546651 1 glog.go:58] Output: Rounding up size to full physical extent 12.00 MiB pmem-csi-node-6crnt/pmem-driver: WARNING: Logical volume ndbus0region0fsdax/5926bc3d-a1aa-11e9-bb06-ae52abbc8405 not zeroed. pmem-csi-node-6crnt/pmem-driver: Logical volume "5926bc3d-a1aa-11e9-bb06-ae52abbc8405" created. pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.546696 1 glog.go:58] Executing: lvs --noheadings --nosuffix -o lv_name,lv_path,lv_size --units B ndbus0region0fsdax pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.603070 1 glog.go:58] Output: 5926bc3d-a1aa-11e9-bb06-ae52abbc8405 /dev/ndbus0region0fsdax/5926bc3d-a1aa-11e9-bb06-ae52abbc8405 12582912 pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.603128 1 glog.go:58] ClearDevice: path: /dev/ndbus0region0fsdax/5926bc3d-a1aa-11e9-bb06-ae52abbc8405 flush:false pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.603155 1 glog.go:58] Zeroing 4 1k blocks at start of device: /dev/ndbus0region0fsdax/5926bc3d-a1aa-11e9-bb06-ae52abbc8405 Size 12582912 pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.603172 1 glog.go:58] Executing: dd if=/dev/zero of=/dev/ndbus0region0fsdax/5926bc3d-a1aa-11e9-bb06-ae52abbc8405 bs=1024 count=4 pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.605697 1 glog.go:58] Output: 4+0 records in pmem-csi-node-6crnt/pmem-driver: 4+0 records out pmem-csi-node-6crnt/pmem-driver: 4096 bytes (4.1 kB, 4.0 KiB) copied, 0.00013222 s, 31.0 MB/s pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.605738 1 glog.go:58] CreateVolume: Record new volume as {5926bc3d-a1aa-11e9-bb06-ae52abbc8405 11534336 map[Name:state-volume-8DA0F868-163CAD76]} pmem-csi-controller-0/pmem-driver: I0708 18:01:02.608769 1 glog.go:58] Chosen nodes: map[XXXXX:1] pmem-csi-controller-0/pmem-driver: I0708 18:01:02.608821 1 glog.go:58] CreateVolume: Record new volume as {5926bc3d-a1aa-11e9-bb06-ae52abbc8405 state-volume-8DA0F868-163CAD76 11534336 map[XXXXX:1]} pmem-csi-controller-0/pmem-driver: I0708 18:01:02.608862 1 glog.go:58] GRPC response: {"volume":{"accessible_topology":[{"segments":{"pmem-csi.intel.com/node":"XXXXX"}}],"capacity_bytes":11534336,"volume_id":"5926bc3d-a1aa-11e9-bb06-ae52abbc8405"}} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.605777 1 glog.go:58] GRPC response: {"volume":{"accessible_topology":[{"segments":{"pmem-csi.intel.com/node":"XXXXXX"}}],"capacity_bytes":11534336,"volume_id":"5926bc3d-a1aa-11e9-bb06-ae52abbc8405"}} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.612409 1 glog.go:58] GRPC call: /csi.v1.Controller/ListVolumes pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.612437 1 glog.go:58] GRPC request: {} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.613429 1 glog.go:58] ListVolumes pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.613447 1 glog.go:58] GRPC response: {"entries":[{"volume":{"capacity_bytes":11534336,"volume_context":{"Name":"state-volume-8DA0F868-163CAD76"},"volume_id":"5926bc3d-a1aa-11e9-bb06-ae52abbc8405"}}]} pmem-csi-controller-0/pmem-driver: I0708 18:01:02.617922 1 glog.go:58] GRPC call: /csi.v1.Controller/GetCapacity pmem-csi-controller-0/pmem-driver: I0708 18:01:02.617949 1 glog.go:58] GRPC request: {} pmem-csi-controller-0/pmem-driver: I0708 18:01:02.619145 1 glog.go:58] Connecting to node controller : tcp://NN.NNN.NN.NNN:10001 pmem-csi-controller-0/pmem-driver: I0708 18:01:02.619167 1 glog.go:58] Connecting to NN.NNN.NN.NN:10001 pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.626192 1 glog.go:58] GRPC call: /csi.v1.Controller/GetCapacity pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.626235 1 glog.go:58] GRPC request: {} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.627174 1 glog.go:58] Executing: vgs --noheadings --nosuffix -o vg_name,vg_size,vg_free,vg_tags --units B ndbus0region1fsdax ndbus0region0fsdax pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.683289 1 glog.go:58] Output: ndbus0region0fsdax 266283778048 266271195136 fsdax pmem-csi-node-6crnt/pmem-driver: ndbus0region1fsdax 266283778048 266283778048 fsdax pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.683341 1 glog.go:58] Executing: vgs --noheadings --nosuffix -o vg_name,vg_size,vg_free,vg_tags --units B ndbus0region1fsdax ndbus0region0fsdax pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.727168 1 glog.go:58] Output: ndbus0region0fsdax 266283778048 266271195136 fsdax pmem-csi-node-6crnt/pmem-driver: ndbus0region1fsdax 266283778048 266283778048 fsdax pmem-csi-controller-0/pmem-driver: I0708 18:01:02.729016 1 glog.go:58] GRPC response: {"available_capacity":266283778048} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.727217 1 glog.go:58] GRPC response: {"available_capacity":266283778048} Jul 8 21:01:02.732: INFO: node "XXXXX" not one of the expected QEMU nodes (worker)) [AfterEach] pmem csi/home/okartau/go/src/github.com/intel/pmem-csi/test/e2e/storage/sanity.go:175 cleanup: deleting state-volume-8DA0F868-163CAD76 = 5926bc3d-a1aa-11e9-bb06-ae52abbc8405 pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.733787 1 glog.go:58] GRPC call: /csi.v1.Node/NodeUnpublishVolume pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.733814 1 glog.go:58] GRPC request: {"target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pmem-sanity-target.LAPJEf","volume_id":"5926bc3d-a1aa-11e9-bb06-ae52abbc8405"} pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.734793 1 glog.go:58] NodeUnpublishVolume: unmount /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pmem-sanity-target.LAPJEf pmem-csi-node-6crnt/pmem-driver: I0708 18:01:02.740729 `

**Comment by okartau** [comment]
> those logs are not run using these lates changes

my bad. I just rebuilt containers without even thinking. The change is on tester side of course. re-testing...

**Comment by okartau** [comment]
no volume shred attempted with changed tester code, works as expected.

File: test/e2e/storage/sanity.go (hunk 2 above)
Comment (avalluri) [reply test/e2e/storage/sanity.go]: Did the requested change.

**Comment by avalluri** [comment]
> Your last commit doesn't explain why this change is necessary. Eventually it'll have to go upstream to reviewers who may have no context for the commit, so can you summarize the issue in the commit message?

Added the context to commit message.

**Comment by avalluri** [comment]
> no volume shred attempted with changed tester code, works as expected.

Thanks @okartau.

**Comment by pohly** [comment]
> Added the context to commit message.

Still not using a spell-checker when writing commit messages? :-} "unpblish" -> "unpublish". But let's not bother fixing that and merge it.
